theano/gof/tests/test_wrapper.py

```diff
@@ -0,0 +1,245 @@
+from theano.tensor import TensorType
+from theano.gof.wrapper import Wrapper, Wrap
+from theano import config
+from theano import tensor
+from theano.tests import unittest_tools as utt
+
+dtype = config.floatX
+tensor_type_0d = TensorType(dtype, tuple())
+scalar_type = Scalar(dtype)
+generic_type = Generic()
+
+
+# A test op to compute `y = a*x^2 + bx + c` for any tensor x, with a, b, c as op params.
+class QuadraticOpFunc(Op):
+    __props__ = ('a', 'b', 'c')
+    params_type = Wrapper(a=tensor_type_0d,
+                          b=scalar_type,
+                          c=generic_type)
+
+    def __init__(self, a, b, c):
+        self.a = a
+        self.b = b
+        self.c = c
+
+    def make_node(self, x):
+        x = tensor.as_tensor_variable(x)
+        return Apply(self, [x], [x.type()])
+
+    def perform(self, node, inputs, output_storage, coefficients):
+        x = inputs[0]
+        y = output_storage[0]
+        y[0] = coefficients.a * (x**2) + coefficients.b * x + coefficients.c
+
+    def c_code_cache_version(self):
+        return (1, 3)
+
+    def c_support_code_apply(self, node, name):
+        float_type = node.inputs[0].type.dtype_specs()[1]
+        return """
+        /* Computes: x = a*x*x + b*x + c for x in tensor. */
+        int quadratic_%(float_type)s(PyArrayObject* tensor, %(float_type)s a, %(float_type)s b, %(float_type)s c) {
+            NpyIter* iterator = NpyIter_New(tensor,
+                NPY_ITER_READWRITE | NPY_ITER_EXTERNAL_LOOP | NPY_ITER_REFS_OK,
+                NPY_KEEPORDER, NPY_NO_CASTING, NULL);
+            if(iterator == NULL) {
+                PyErr_SetString(PyExc_RuntimeError, "Unable to iterate over a tensor for an elemwise operation.");
+                return -1;
+            }
+            NpyIter_IterNextFunc* get_next = NpyIter_GetIterNext(iterator, NULL);
+            char** data_ptr = NpyIter_GetDataPtrArray(iterator);
+            npy_intp* stride_ptr = NpyIter_GetInnerStrideArray(iterator);
+            npy_intp* innersize_ptr = NpyIter_GetInnerLoopSizePtr(iterator);
+            do {
+                char* data = *data_ptr;
+                npy_intp stride = *stride_ptr;
+                npy_intp count = *innersize_ptr;
+                while(count) {
+                    %(float_type)s x = *((%(float_type)s*)data);
+                    *((%(float_type)s*)data) = a*x*x + b*x + c;
+                    data += stride;
+                    --count;
+                }
+            } while(get_next(iterator));
+            NpyIter_Deallocate(iterator);
+            return 0;
+        }
+        """ % {'float_type': float_type}
+
+    def c_code(self, node, name, inputs, outputs, sub):
+        X = inputs[0]
+        Y = outputs[0]
+        coeff = sub['params']
+        fail = sub['fail']
+        float_type = node.inputs[0].type.dtype_specs()[1]
+        float_typenum = numpy.dtype(node.inputs[0].type.dtype).num
+        coeff_type = 'npy_' + numpy.dtype(dtype).name
```

Comment:
This is a bad example. You should find the dtype without access to a global variable. You should look in node.params_type